ompi/mca/hook/comm_method/hook_comm_method_fns.c

```diff
@@ -0,0 +1,788 @@
+/*
+ * Copyright (c) 2016-2018 IBM Corporation. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+
+#include "hook_comm_method.h"
+
+#ifdef HAVE_DLFCN_H
+#include <dlfcn.h>
+#endif
+
+#include "ompi/communicator/communicator.h"
+#include "ompi/mca/pml/pml.h"
+#include "opal/mca/btl/btl.h"
+#include "ompi/mca/pml/base/base.h"
+#include "ompi/mca/bml/base/base.h"
+#include "ompi/mca/mtl/base/base.h"
+
+// return the pml's module:component:name function pointer in fp
+static char*
+lookup_pml_name(void)
+{
+    return (char*) mca_pml_base_selected_component.pmlm_version.mca_component_name;
+}
+// return the mtl's module:component:name function pointer in fp
+static char*
+lookup_mtl_name(void)
+{
+    if (!ompi_mtl_base_selected_component) { return NULL; }
+    return (char*) ompi_mtl_base_selected_component->mtl_version.mca_component_name;
+}
+// Find the send btl's module:component:name for the incoming comm,rank
+static char*
+lookup_btl_name_for_send(ompi_communicator_t* comm, int rank) {
+    ompi_proc_t *dst_proc = ompi_comm_peer_lookup (comm, rank);
+
+    mca_bml_base_endpoint_t* endpoint = mca_bml_base_get_endpoint(dst_proc);
+    if (endpoint &&
+        endpoint->btl_send.bml_btls &&
+        endpoint->btl_send.bml_btls[0].btl)
+    {
+        return (char*)
+            endpoint->btl_send.bml_btls[0].btl->btl_component->btl_version.mca_component_name;
+    }
+    return NULL;
+}
+
+char *strings_map[] = {
```

Comment:
This approach is extremely static. Why not using a hash table for the names ?

Reply:
I guess it could be a hash, but I'd still need a way to convert the hashes back to strings. So I'd either still have a static list of strings to walk or else I'd have all the ranks including their list of strings in the messages they send to rank 0 so it could have a complete list of strings.

The other part is right now the code can distinguish between "ofi" as an MTL and "ofi" as a BTL. If it was just a general string hash I don't think it would know that both existed. Maybe that's a minor enough feature I shouldn't care.

I'd hate for the table that currently says "tcp" to get unnecessarily bloated to "tcp(btl)" etc.

If we throw out the first long-format table
```
>    host | 0    1    2    3    4    5    6    7    8
>   ======|==============================================
>       0 : sm   tcp  tcp  tcp  tcp  tcp  tcp  tcp  tcp
>       1 : tcp  sm   tcp  tcp  tcp  tcp  tcp  tcp  tcp
>       2 : tcp  tcp  self tcp  tcp  tcp  tcp  tcp  tcp
>       3 : tcp  tcp  tcp  self tcp  tcp  tcp  tcp  tcp
>       4 : tcp  tcp  tcp  tcp  self tcp  tcp  tcp  tcp
>       5 : tcp  tcp  tcp  tcp  tcp  self tcp  tcp  tcp
>       6 : tcp  tcp  tcp  tcp  tcp  tcp  self tcp  tcp
>       7 : tcp  tcp  tcp  tcp  tcp  tcp  tcp  self tcp
>       8 : tcp  tcp  tcp  tcp  tcp  tcp  tcp  tcp  self
```
and only used the more abbreviated form
```
>    host | 0 1 2 3 4       8
>   ======|====================
>       0 : A C C C C C C C C
>       1 : C A C C C C C C C
>       2 : C C B C C C C C C
>       3 : C C C B C C C C C
>       4 : C C C C B C C C C
>       5 : C C C C C B C C C
>       6 : C C C C C C B C C
>       7 : C C C C C C C B C
>       8 : C C C C C C C C B
>   key: A == sm
>   key: B == self
>   key: C == tcp
```
then the extra characters wouldn't give me any heartburn. But I still find the long form more readable than the above so prefer it on the smaller runs

Reply:
Can you comment why some hosts talk to the same host over sm, and other hosts use self?
Can we infer that the hosts that talk to themselves over self only have 1 rank on them?

Reply:
Yeah, it's identifying the local-host traffic.  If there's one rank on the host then it's just 'self'. If there are multiple ranks then both 'self' and 'vader' would be active between ranks on that host, so it shows 'vader' as that's the more useful piece of info

Reply:
I have to say the reading of such table is not trivial, as one might expect to read connectivity between MPI ranks when in fact this table shows connectivity (one of the possible BTL) between nodes (all ranks accumulated together). Moreover, it works best for regular and symmetric cases on homogeneous environments, the exact places where the best answer would have been "tcp across nodes, vader between processes on the same node, and self for communication in the same process".